File: implementations/elixir/lib/channel/cipher_state.ex
defmodule Ockam.Channel.CipherState do
  alias Ockam.Vault
  alias Ockam.Vault.Secret

  @type cipher :: :aes_256_gcm | :chachapoly
  @type nonce :: non_neg_integer()
  @type key :: Secret.t() | nil

  @ciphers [:aes_256_gcm]

  defstruct k: nil, n: 0, cipher: :aes_256_gcm

  @type t :: %__MODULE__{
          k: key(),
          n: nonce(),
          cipher: :aes_256_gcm
        }

  def init(cipher), do: init(nil, cipher)

  def init(key, cipher) when cipher in @ciphers do
    do_init(key, cipher)
  end

  defp do_init(%Secret{} = key, cipher), do: %__MODULE__{k: key, n: 0, cipher: cipher}
  defp do_init(nil, cipher), do: %__MODULE__{k: nil, n: 0, cipher: cipher}

  def set_key(%__MODULE__{} = state, %Secret{} = key) do
    %__MODULE__{state | k: key, n: 0}
  end

  def has_key(%__MODULE__{k: k}), do: k != nil

  def set_nonce(%__MODULE__{} = state, nonce) do
    %__MODULE__{state | n: nonce}
  end

  def encrypt(%__MODULE__{k: nil} = state, _vault, _aad, plaintext) do
    {:ok, state, plaintext}
  end

  def encrypt(%__MODULE__{k: k, n: n, cipher: :aes_256_gcm} = state, vault, aad, plaintext) do
    with {:ok, ciphertext} <- Vault.encrypt(vault, k, n, aad, plaintext) do
      {:ok, %__MODULE__{state | n: n + 1}, ciphertext}
    end
  end

  def decrypt(%__MODULE__{k: nil} = state, _vault, _aad, ciphertext) do
    {:ok, state, ciphertext}
  end

  def decrypt(%__MODULE__{k: k, n: n, cipher: :aes_256_gcm} = state, vault, aad, ciphertext) do
    with {:ok, plaintext} <- Vault.decrypt(vault, k, n, aad, ciphertext) do
      {:ok, %__MODULE__{state | n: n + 1}, plaintext}
    end
  end

  def rekey(%__MODULE__{k: k, cipher: :aes_256_gcm} = cs, vault) do
    {:ok, k} = Vault.rekey(vault, k)
    %__MODULE__{cs | k: k}
  end

  def cipher(%__MODULE__{cipher: cipher}), do: cipher
  def key(%__MODULE__{k: k}), do: k
end

File: implementations/rust/ockam/ockam_key_exchange_xx/src/noise/cipher_state.rs
use crate::{XXError, XXVault};
use ockam_core::vault::{
    Buffer, Secret, SecretAttributes, SecretPersistence, SecretType, AES256_SECRET_LENGTH,
};
use ockam_core::Result;

/// A CipherState object contains k and n variables, which it uses to encrypt
/// and decrypt ciphertexts. During the handshake phase each party has a single CipherState,
/// but during the transport phase each party
/// has two CipherState objects: one for sending, and one for receiving.
#[derive(Debug)]
pub struct CipherState<V: XXVault> {
    k: Option<Secret>,
    n: u64,
    vault: V,
}

impl<V: XXVault> CipherState<V> {
    /// k
    pub fn k(&self) -> Option<Secret> {
        self.k.clone()
    }
    /// n
    pub fn n(&self) -> u64 {
        self.n
    }
}

impl<V: XXVault> CipherState<V> {
    pub(crate) fn new(vault: V, key: Option<Secret>) -> Self {
        let mut res = Self {
            k: None,
            n: 0,
            vault,
        };

        res.initialize_key(key);

        res
    }

    pub(crate) fn convert_nonce(nonce: u64) -> [u8; 12] {
        let mut n: [u8; 12] = [0; 12];
        n[4..].copy_from_slice(&nonce.to_be_bytes());

        n
    }
}

impl<V: XXVault> CipherState<V> {
    /// Sets k = key. Sets n = 0.
    pub fn initialize_key(&mut self, key: Option<Secret>) {
        self.k = key;
        self.n = 0;
    }

    /// Returns true if k is non-empty, false otherwise.
    pub fn has_key(&self) -> bool {
        self.k.is_some()
    }

    /// Sets n = nonce. This function is used for handling out-of-order transport messages,
    /// as described in Section 11.4.
    pub fn set_nonce(&mut self, nonce: u64) {
        self.n = nonce;
    }

    /// If k is non-empty returns ENCRYPT(k, n++, ad, plaintext). Otherwise returns plaintext.
    pub async fn encrypt_with_ad(&mut self, ad: &[u8], plaintext: &[u8]) -> Result<Buffer<u8>> {
        let k;
        if let Some(k_val) = self.k.clone() {
            k = k_val;
        } else {
            return Ok(Buffer::from(plaintext));
        }

        let n = Self::convert_nonce(self.n);
        let res = self
            .vault
            .aead_aes_gcm_encrypt(&k, plaintext, &n, ad)
            .await?;

        self.n += 1;

        Ok(res)
    }

    /// If k is non-empty returns DECRYPT(k, n++, ad, ciphertext).
    /// Otherwise returns ciphertext.
    /// If an authentication failure occurs in DECRYPT() then n is not incremented
    /// and an error is signaled to the caller.
    pub async fn decrypt_with_ad(&mut self, ad: &[u8], ciphertext: &[u8]) -> Result<Buffer<u8>> {
        let k;
        if let Some(k_val) = self.k.clone() {
            k = k_val;
        } else {
            return Ok(Buffer::from(ciphertext));
        }

        let n = Self::convert_nonce(self.n);
        let res = self
            .vault
            .aead_aes_gcm_decrypt(&k, ciphertext, &n, ad)
            .await?;

        self.n += 1;

        Ok(res)
    }

    /// Sets k = REKEY(k).
    ///
    /// Returns a new 32-byte cipher key as a pseudorandom function of k.
    /// If this function is not specifically defined for some set of cipher functions,
    /// then it defaults to returning the first 32 bytes from ENCRYPT(k, maxnonce, zerolen, zeros),
    /// where maxnonce equals 264-1, zerolen is a zero-length byte sequence,
    /// and zeros is a sequence of 32 bytes filled with zeros.
    pub async fn rekey(&mut self) -> Result<()> {
        let k;
        if let Some(k_val) = self.k.clone() {
            k = k_val;
        } else {
            return Err(XXError::InvalidState.into());
        }

        let nonce = Self::convert_nonce(u64::MAX);
        let k = &self
            .vault
            .aead_aes_gcm_encrypt(&k, &[0u8; 32], &nonce, &[])
            .await?[..AES256_SECRET_LENGTH];

        let k = self
            .vault
            .secret_import(
                k,
                SecretAttributes::new(
                    SecretType::Aes,
                    SecretPersistence::Ephemeral,
                    AES256_SECRET_LENGTH,
                ),
            )
            .await?;

        self.k = Some(k);

        Ok(())
    }
}
